stan/math/rev/fun/singular_values.hpp

```diff
@@ -0,0 +1,75 @@
+#ifndef MATH_SINGULAR_VALUES_H
+#define MATH_SINGULAR_VALUES_H
+
+#include <stan/math/rev/meta.hpp>
+#include <stan/math/rev/core.hpp>
+#include <stan/math/rev/functor/reverse_pass_callback.hpp>
+#include <stan/math/rev/core/arena_matrix.hpp>
+#include <stan/math/rev/fun/value_of.hpp>
+#include <stan/math/rev/fun/typedefs.hpp>
+#include <stan/math/prim/fun/Eigen.hpp>
+#include <stan/math/prim/err/check_symmetric.hpp>
+#include <stan/math/prim/err/check_nonzero_size.hpp>
+#include <stan/math/prim/fun/typedefs.hpp>
+#include <stan/math/prim/fun/value_of_rec.hpp>
+#include <Eigen/SVD>
+
+namespace stan {
+namespace math {
+
+/**
+ * Return the singular values of the specified matrix.
+ *
+ * Equation (4) from Differentiable Programming Tensor Networks
+ * is used for MxN input matrix's adjoint calculation.
+ *
+ * @tparam EigMat type of input matrix
+ * @param m MxN input matrix
+ * @return Singular values of matrix
+ */
+template <typename EigMat,
+    require_eigen_matrix_dynamic_t<EigMat>* = nullptr,
+	  require_vt_var<EigMat>* = nullptr>
+inline auto singular_values(const EigMat& m) {
+  using ret_type = promote_scalar_t<var, Eigen::VectorXd>;
+  check_nonzero_size("singular_values", "m", m);
+
+  const int M = to_arena(m.rows());
+  auto arena_m = to_arena(m);
+  
+  Eigen::MatrixXd m_val = value_of(arena_m);
+  Eigen::JacobiSVD<Eigen::MatrixXd> svd(m_val, Eigen::ComputeFullU | Eigen::ComputeFullV);
+
+  auto singular_values = to_arena(svd.singularValues());
+  auto U = to_arena(svd.matrixU());
+  auto V = to_arena(svd.matrixV());
+  Eigen::MatrixXd Fp(M,M);
+  Eigen::MatrixXd Fm(M,M);
+  for(int i = 0; i < M; i++) {
+    for(int j = 0; j < M; j++) {
+      if(j == i) {
+        Fp(i, j) = 0.0;
+        Fm(i, j) = 0.0;
+      } else {
+        Fp(i, j)  = 1.0 / (singular_values[j] - singular_values[i]) + 1.0 / (singular_values[i] + singular_values[j]);
+        Fm(i, j) = 1.0 / (singular_values[j] - singular_values[i]) - 1.0 / (singular_values[j] + singular_values[i]);
+      }
+    }
+  }
+  auto arena_Fp = to_arena(Fp);
+  auto arena_Fm = to_arena(Fm);
+
+  reverse_pass_callback(
```

Comment:
`reverse_pass_callback` creates a `vari` on the `chainable_stack`. When that `vari` 's chain method is called, it calls the lambda defined here.

In this situation, this lambda needs to propagate the adjoints from the output of this function (the singular values) back to the inputs (the matrix m).

So you have to update the adjoint of m from the adjoint of the singular values.

The adjoints of U and V are zero. The way you can see this is: In a Stan model, the adjoint of any variable is the gradient of the log density with respect to that variable. Adjoints start at zero and are incremented in the reverse pass. Because U and V are not returned from this function, nothing increments their adjoints, and so adjU and adjV are zero.

Section 1.1, Figure 2 and 3 explain the setup: https://arxiv.org/pdf/1509.07164.pdf . Figure 2 is the forward pass, and Figure 3 is the reverse pass (each line of the table corresponds to what could be one chain call or one `reverse_pass_callback`)